specs/core/0_beacon-chain.md

```diff
@@ -1739,61 +1691,87 @@ def update_validator_registry(state: BeaconState) -> None:
-
+            exit_validator(state, index)
+            
+            
+    state.validator_registry_latest_change_slot = state.slot
+```
 
-    # Calculate the total ETH that has been penalized in the last ~2-3 withdrawal periods
-    period_index = current_slot // COLLECTIVE_PENALTY_CALCULATION_PERIOD
-    total_penalties = (
-        (latest_penalized_exit_balances[period_index]) +
-        (latest_penalized_exit_balances[period_index - 1] if period_index >= 1 else 0) +
-        (latest_penalized_exit_balances[period_index - 2] if period_index >= 2 else 0)
-    )
+Regardless of whether the above conditions are satisfied, run the following:
 
-    # Calculate penalties for slashed validators
-    def to_penalize(index):
-        return state.validator_registry[index].status == EXITED_WITH_PENALTY
-    validators_to_penalize = filter(to_penalize, range(len(validator_registry)))
-    for index in validators_to_penalize:
-        state.validator_balances[index] -= get_effective_balance(state, index) * min(total_penalties * 3, total_balance) // total_balance
+```python
+def process_penalties_and_exits(state: BeaconState) -> None:
+    # The active validators
+    active_validator_indices = get_active_validator_indices(state.validator_registry, state.slot)
+    # The total effective balance of active validators
+    total_balance = sum([get_effective_balance(state, i) for i in active_validator_indices])
 
-    return validator_registry, latest_penalized_exit_balances, validator_registry_delta_chain_tip
+    for i, validator in enumerate(validators):
+        if (state.slot // EPOCH_LENGTH) - (validator.penalized_slot // EPOCH_LENGTH) == LATEST_PENALIZED_EXIT_LENGTH // 2:
+            e = (state.slot // EPOCH_LENGTH) % LATEST_PENALIZED_EXIT_LENGTH
+            total_at_start = state.latest_penalized_exit_balances[(e + 1) % LATEST_PENALIZED_EXIT_LENGTH]
+            total_at_end = state.latest_penalized_exit_balances[e]
+            total_penalties = total_at_end - total_at_start
+            penalty = get_effective_balance(state, i) * min(total_penalties * 3, total_balance) // total_balance
+            state.validator_balances[index] -= penalty
+
+    def eligible(index):
+        validator = state.validator_registry[index]
+        if validator.penalized_slot <= state.slot:
+            PENALIZED_WITHDRAWAL_TIME = LATEST_PENALIZED_EXIT_LENGTH * EPOCH_LENGTH // 2
+            return state.slot >= validator.penalized_slot + PENALIZED_WITHDRAWAL_TIME
+        else:
+            return state.slot >= validator.exit_slot + MIN_VALIDATOR_WITHDRAWAL_TIME
+
+    eligible_indices = filter(eligible, all_indices)
+    sorted_indices = sorted(eligible_indices, filter=lambda index: state.validator_registry[index].exit_count)
+    withdrawn_so_far = 0
+    for index in sorted_indices:
+        validator = state.validator_registry[index]
+        if validator.penalized_slot <= state.slot:
```

Comment:
We can remove this conditional, right? We are already doing penalizations above.

Reply:
Done!